src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala

```diff
@@ -304,54 +716,292 @@ class MaxPooling(val param:LayerParameter, val id:Int, val net:CaffeNetwork) ext
+   *  - db: Gradient wrt `b`, of shape (F, 1).
+   */
   override def backward(dmlScript:StringBuilder, outSuffix:String) = 
-    invokeBackward(dmlScript, outSuffix, List[String](dX, dWeight, dBias), dout, Hout, Wout, X, weight, bias, numChannels, Hin, Win, kernel_h, kernel_w, stride_h, stride_w, pad_h, pad_w)
+    invokeBackward(dmlScript, outSuffix, List[String]("dOut" + id, dWeight, dBias), dout, Hout, Wout, X, weight, bias, numChannels, Hin, Win, kernel_h, kernel_w, stride_h, stride_w, pad_h, pad_w)
+  // n * c_o * h_o * w_o, where h_o = (h_i + 2 * pad_h - kernel_h) / stride_h + 1 and w_o likewise.
   override def outputShape = ( numKernels, Hout, Wout )
   // -------------------------------------------------
   def numChannels = bottomLayerOutputShape._1
   def Hin = bottomLayerOutputShape._2
   def Win = bottomLayerOutputShape._3
   def Hout = ConvolutionUtils.getConv2dOutputMap(bottomLayerOutputShape._2, kernel_h, stride_h, pad_h) 
   def Wout =  ConvolutionUtils.getConv2dOutputMap(bottomLayerOutputShape._3, kernel_w, stride_w, pad_w)
+  // -------------------------------------------------
+  def convParam = param.getConvolutionParam
+  // num_output (c_o): the number of filters
+  def numKernels = convParam.getNumOutput.toString
+  // kernel_size (or kernel_h and kernel_w): specifies height and width of each filter
+  def kernel_h = if(convParam.hasKernelH) convParam.getKernelH.toString 
+                   else if(convParam.getKernelSizeCount > 0)  convParam.getKernelSize(0).toString 
+                   else throw new LanguageException("Incorrect kernel parameters")
+  def kernel_w = if(convParam.hasKernelW) convParam.getKernelW.toString 
+                   else if(convParam.getKernelSizeCount > 0)  convParam.getKernelSize(0).toString 
+                   else throw new LanguageException("Incorrect kernel parameters")
+  // stride (or stride_h and stride_w) [default 1]: specifies the intervals at which to apply the filters to the input
+  def stride_h = if(convParam.hasStrideH) convParam.getStrideH.toString 
+                   else if(convParam.getStrideCount > 0)  convParam.getStride(0).toString 
+                   else "1"
+  def stride_w = if(convParam.hasStrideW) convParam.getStrideW.toString 
+                   else if(convParam.getStrideCount > 0)  convParam.getStride(0).toString 
+                   else "1"
+  // pad (or pad_h and pad_w) [default 0]: specifies the number of pixels to (implicitly) add to each side of the input
+  def pad_h =   if(convParam.hasPadH) convParam.getPadH.toString 
+                   else if(convParam.getPadCount > 0)  convParam.getPad(0).toString 
+                   else "0"
+  def pad_w =   if(convParam.hasPadW) convParam.getPadW.toString 
+                   else if(convParam.getPadCount > 0)  convParam.getPad(0).toString 
+                   else "0"
+}
+
+class DeConvolution(val param:LayerParameter, val id:Int, val net:CaffeNetwork) extends CaffeLayer with HasWeight with HasBias {
```

Comment:
What do you think about updating this to, say, `Convolution2DTranspose`?  I know that Caffe used "deconvolution", but it would be nice to not perpetuate that incorrect naming.  Also, Caffe2 now uses "ConvTranspose" instead of "Deconvolution" --> https://caffe2.ai/docs/operators-catalogue.html#convtranspose.